Restore API + move to node

remy opened the pull request:
Looks like I was able to keep your commits (though I tossed the typescript - sorry, the core of all the works is based on your stuff @Richienb).

Imported from https://github.com/remy/mit-license/pull/1348: Closes #1286, Closes #1299, Closes #765, Closes #659, Closes #73.
Since we're migrating to Node: Closes #764, Closes #806, Closes #830, Closes #904, Closes #1320

Richienb commented:
@remy There seems to still be clutter from removing TypeScript which includes leftover dependencies.

Also, it seems that the package manager was changed back to NPM.
To avoid a [ReDOS issue](https://npmjs.com/advisories/786), it needs to be switched back to Yarn since it can override the vulnerable version with a patched version fetched from the `package.json` file:
https://github.com/remy/mit-license/blob/d5636b1004ff78a1d3ee30ac2f1641a00d9ebc54/package.json#L55-L57

Travis CI somehow upset itself during the rewrite due to nvm version pinning and the package manager change.

I'd like to fix these issues but as the PR branch is in this repo, I don't have collaborator permissions.

remy commented:
Added you as a collaborator 👍

I moved it to npm only because of that Android browser error that we saw during the twitch session - there's a bunch of threads on it as it appears to be a new issue (and one solution was to move to npm, but looking back this didn't fix it - so happy for you to switch back to yarn) - the solution was to pin `caniuse-lite@1.0.30000974`

On deps - yep, I can see that too - go ahead and remove as you feel fit.

I also removed some packages that I didn't think were needed, one such was `dayjs` was being used to get the current year - we're able to get that from straight from JavaScript via `new Date().getFullYear()` and I couldn't see that it was used elsewhere.

You'll see but I added some sanitisation too which I don't think was ever around in the PHP version.

Otherwise - it was a really good PR (particularly the restoring of the POST support) - and I reckon this is fairly close to done with a few tweaks. Looking forward to seeing your changes too 👍

Richienb commented:
@remy The `.nvmrc` file seems to do the same thing as `engines.node` property in the `package.json` file to pin the version.

After some research, we technically only need to keep one of these.

remy commented:
@Richienb happy to drop it (and add it as an exclusion on gitignore). I use it in my terminal so that when I drop into that directory, I auto switch to the version of node in the `.nvmrc` file. Just a shame I can't work out how to get my terminal to do the same thing from `engines.node`.

remy commented:
Oh, I think I committed a change that commented out the CSS post processing. That wasn't intentional - it was just to speed up my development workflow (it was adding a few seconds to node starting up - no issue in production, but slow for the code/reload/retest cycle).

remy commented on .editorconfig:
```diff
@@ -0,0 +1,14 @@
+trim_trailing_whitespace = true
+insert_final_newline = true
+indent_style = space
+indent_size = 2
```
Sorry! 😄

remy commented on .gitignore:
```diff
@@ -128,91 +45,12 @@ Network Trash Folder
-build/Release
-
-# Dependency directories
-node_modules/
```
Probably want this (and `.env`) back in the ignore file

Richienb replied:
Yup. I somehow forgot to readd all the node gitignore rules.

remy commented on server.js:
```diff
@@ -57,7 +62,20 @@ app.use(require('./middleware/load-options'));
-  console.log(`🚀 on http://localhost:${PORT}`);
-});
+// If the current process is the main one
+if (cluster.isMaster) {
```
Can you revert this change. If we want to merge this code, we need to do it in small succinct parts. I don't want to add multi-core it adds more plates and the production instance is single core in the first place.

Richienb replied:
@remy I've rolled it back.

remy commented on routes/get.js:
```diff
@@ -48,7 +48,7 @@ module.exports = (req, res) => {
       user.email.trim().toLowerCase()
     )}" />`;
-  } else if (typeof user.copyright[0] === 'object' && user.gravatar) {
+  } else if (user.copyright && typeof user.copyright[0] === 'object' && user.gravatar) {
```
I thought this could never happen because of the middleware I put in place? Just curious.

Richienb replied:
Well, it actually did happen.

The line change solved the following error when calling the root domain:
```
Uncaught TypeError: Cannot read property '0' of undefined
```

remy commented on middleware/load-user.js:
```diff
@@ -14,21 +14,13 @@ module.exports = async (req, res, next) => {
+    .catch(({code, message}) => {
+      if (code !== 'ENOENT') res.code(500).send(`An internal error occurred - open an issue on https://github.com/remy/mit-license with the following information: ${message}`)
+    })
+    .finally(() => next())
```
I'm pretty sure changing this to promises with `finally` is obscuring an error. I'm not 100% sure, but can you tell me what happens when the catch fires, and `res.send` is called and then finally `next()` is called? 

My guess: headers already sent - though I'm not 100% sure.

Richienb commented:
@remy It's an [anti-pattern](https://dev.to/maxart2501/gotchas-about-asyncawait-and-promises-9di) solved by switching try/catch to then/catch/finally.

remy commented:
@Richienb what's the anti-pattern? It's morning for me (6am) and that article is pretty long. What is it specifically?

Richienb commented:
@remy In the article, it's documented under the header **Scoped woes**.

remy commented:
👍

Probably worth my looking at why this tripped the code up (the middleware
default was supposed to preempt that specific bug)

On Fri, 21 Jun 2019, 06:52 Richie Bendall, <notifications@github.com> wrote:

> *@Richienb* commented on this pull request.
> ------------------------------
>
> In routes/get.js
> <https://github.com/remy/mit-license/pull/1355#discussion_r296104025>:
>
> > @@ -48,7 +48,7 @@ module.exports = (req, res) => {
>      gravatar = `<img id="gravatar" alt="Profile image" src="https://www.gravatar.com/avatar/${md5(
>        user.email.trim().toLowerCase()
>      )}" />`;
> -  } else if (typeof user.copyright[0] === 'object' && user.gravatar) {
> +  } else if (user.copyright && typeof user.copyright[0] === 'object' && user.gravatar) {
>
> Well, it actually did happen.
>
> The line change solved the following error when calling the root domain:
>
> Uncaught TypeError: Cannot read property '0' of undefined
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/remy/mit-license/pull/1355?email_source=notifications&email_token=AAADLBDIA2TANOMBQSHYSSLP3RUCDA5CNFSM4HZLVZCKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB4HSJ5Y#discussion_r296104025>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAADLBHIBG2PI3T3WX4CEE3P3RUCDANCNFSM4HZLVZCA>
> .
>

remy commented:
On anti pattern, reading that section, it's not an anti pattern, it's just the author thinks it's ugly because the variable is declared outside of the try scope. How code looks is mostly a matter of preference (I like 2 spaces, you like 4, one is ugly looking to the other!).

Richienb commented:
@remy I'll revert back to try/catch

Richienb commented:
@remy I figured out the problem with the middleware.

At first, the object is set to:
```json
{
    "copyright": "<copyright holders>"
}
```
But when the user file is loaded, it is overwritten instead of appended to. I'll commit a fix soon.

Richienb commented:
@remy I believe this PR is now complete.

Richienb commented:
@remy I believe we can now have a look at merging this PR.

Richienb commented:
In the last commit, I updated `caniuse-lite` to the latest version which seems to have fixed the browserlist error: `BrowserslistError: Unknown browser query android all. Maybe you are using old Browserslist or made typo in query.`.

However, if the error continues, readd `"caniuse-lite": "1.0.30000974"` to the `resolutions` section of the `package.json` and rerun `yarn install --force`.

remy commented:
Noted 👌

I'm just embarking on a family holiday and when I'm back late next week I'll get this merged and update the production environment so that it uses node (and not PHP).
